modules/core/src/main/java/org/apache/ignite/internal/util/collection/IntHashMap.java

```diff
@@ -0,0 +1,337 @@
+
+                    return tmp;
+                }
+
+                entries[curIdx] = entries[nextIdx];
+            }
+
+            throw new IllegalStateException("Unreachable state exception. Backward shift has a problem. " +
+                "Removing key: " + key + " map state: " + toString());
+        }
+        finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public V putIfAbsent(int key, V val) {
+        lock.writeLock().lock();
+        try {
+            int idx = find(key);
+
+            if (idx < 0) {
+                put(key, val);
+
+                return null;
+            }
+            else
+                return entries[idx].val;
+        }
+        finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public <E extends Throwable> void forEach(EntryConsumer<V, E> act) throws E {
+        lock.readLock().lock();
+        try {
+            for (Entry<V> entry : entries)
+                if (entry != null)
+                    act.accept(entry.key, entry.val);
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public int size() {
+        lock.readLock().lock();
+        try {
+            return size;
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean containsKey(int key) {
+        lock.readLock().lock();
+        try {
+            return find(key) >= 0;
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean containsValue(V val) {
+        lock.readLock().lock();
+        try {
+            return Arrays.stream(entries)
+                .filter(Objects::nonNull)
+                .anyMatch(entry -> Objects.equals(val, entry.val));
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString() {
+        lock.readLock().lock();
+        try {
+            String strEntries = Arrays.stream(entries)
+                .filter(Objects::nonNull)
+                .map(Entry::toString)
+                .collect(Collectors.joining(","));
+
+            return "IntHashMap{size=" + size + ", entries=[" + strEntries + "]}";
+        }
+        finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Returns size of shift from ideal position computing via index(key).
+     *
+     * @param curIdx Current index.
+     * @param key Target key.
+     */
+    protected int distance(int curIdx, int key) {
+        int keyIdx = index(key);
+
+        return curIdx >= keyIdx ? curIdx - keyIdx : entries.length - keyIdx + curIdx;
```

Comment:
"curIdx >= keyIdx" uses subtraction implicitly, so you make it twice.
"curIdx - keyIdx" can be calculated a line before, comparing it to zero will be "much faster".